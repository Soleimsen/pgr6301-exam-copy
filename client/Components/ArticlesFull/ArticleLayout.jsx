import React from "react";
import { Link } from "react-router-dom";

const ArticleLayout = ({
  article: { _id, author, date, headline, body, tags },
}) => {
  const dateFormat = new Date(date).toLocaleDateString();
  return (
    <>
      <div key={_id} className="articleLayout">
        <h1>{headline}</h1>
        <div className="dateName">
          <p>
            <i>Author: {author}</i>
          </p>
          <p className="date">Date written: {dateFormat}</p>
        </div>
        <div className="border">
          <div>
            <p className="articleBody">{body}</p>
            <p>Topic: {tags}</p>
            <Link className="articleBody" to={`/articles/${_id}`}>
              Read Article
            </Link>
          </div>
        </div>
      </div>
    </>
  );
};

export default ArticleLayout;
