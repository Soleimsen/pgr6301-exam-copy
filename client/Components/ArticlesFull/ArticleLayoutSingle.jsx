import React from "react";
import { Link } from "react-router-dom";

const ArticleLayoutSingle = ({
  user,
  article: { _id, author, date, headline, body, tags, sub },
}) => {
  const dateFormat = new Date(date).toLocaleDateString();

  return (
    <>
      <div key={_id} className="articleLayout">
        <h1>{headline}</h1>
        <div className="dateName">
          <p>
            <i>Author: {author}</i>
          </p>
          <p className="date">Date written: {dateFormat}</p>
        </div>
        <div className="border">
          <div>
            <p className="articleBody">{body}</p>
            <p>Topic: {tags}</p>
            {
              /*sub === user.azure.sub && (*/
              <Link to={`/articles/edit/${_id}`}>
                Edit Article(bare klikk om du er logget inn med AD. rakk ikke
                fikse)
              </Link>
              /*)*/
            }
          </div>
        </div>
      </div>
    </>
  );
};

export default ArticleLayoutSingle;
