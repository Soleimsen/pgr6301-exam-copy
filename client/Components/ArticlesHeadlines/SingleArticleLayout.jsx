import React from "react";
import { Link } from "react-router-dom";

const SingleArticleLayout = ({
  article: { _id, author, date, headline, body },
}) => {
  const dateFormat = new Date(date).toLocaleDateString();
  return (
    <>
      <Link className="articleLayout" key={_id} to={`/articles/${_id}`}>
        <div className="border">
          <h1>{headline}</h1>
          <div className="dateName">
            <p>
              <i>Author: {author}</i>
            </p>
            <p className="date">Date written: {dateFormat}</p>
          </div>
        </div>
      </Link>
    </>
  );
};

export default SingleArticleLayout;
