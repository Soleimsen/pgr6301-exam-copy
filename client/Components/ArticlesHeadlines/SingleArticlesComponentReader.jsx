import React from "react";
import { UseLoading } from "../../Hooks/UseLoading";
import { fetchJSON } from "../../Hooks/FetchJSON";
import SingleArticleLayoutReader from "./SingleArticleLayoutReader";

const SingleArticlesComponentReader = () => {
  const { loading, error, data } = UseLoading(async () =>
    fetchJSON("/api/articles")
  );

  if (loading) {
    return (
      <div>
        <div>Loading...</div>
      </div>
    );
  }
  if (error) {
    return (
      <div>
        <h1>Error</h1>
        <div>{error.toString()}</div>
      </div>
    );
  }
  return (
    <>
      <div className="articlesContainer">
        <h1>Articles</h1>
        <ul className="article">
          {data
            .slice()
            .reverse()
            .map((article) => (
              <SingleArticleLayoutReader key={article._id} article={article} />
            ))}
        </ul>
      </div>
    </>
  );
};

export default SingleArticlesComponentReader;
