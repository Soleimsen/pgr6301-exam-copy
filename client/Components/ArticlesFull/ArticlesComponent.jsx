import React from "react";
import { UseLoading } from "../../Hooks/UseLoading";
import { fetchJSON } from "../../Hooks/FetchJSON";
import ArticleLayout from "./ArticleLayout";

const ArticlesComponent = ({ d }) => {
  const { loading, error, data } = UseLoading(async () =>
    fetchJSON("/api/articles")
  );

  if (loading) {
    return (
      <div>
        <div>Loading...</div>
      </div>
    );
  }
  if (error) {
    return (
      <div>
        <h1>Error</h1>
        <div>{error.toString()}</div>
      </div>
    );
  }
  return (
    <>
      <div className="articlesContainer">
        <h1>Articles</h1>
        <ul className="article">
          {data
            .slice()
            .reverse()
            .map((article) => (
              <ArticleLayout key={article._id} article={article} />
            ))}
        </ul>
      </div>
    </>
  );
};

export default ArticlesComponent;
