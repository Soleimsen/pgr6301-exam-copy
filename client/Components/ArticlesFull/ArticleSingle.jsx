import React from "react";
import { UseLoading } from "../../Hooks/UseLoading";
import { fetchJSON } from "../../Hooks/FetchJSON";
import { useParams } from "react-router-dom";
import { useArticle } from "../useArticle";
import ArticleLayoutSingle from "./ArticleLayoutSingle";
import { Link } from "react-router-dom";

const ArticleSingle = ({ user }) => {
  const { _id } = useParams();
  const article = useArticle(_id);
  console.log(_id);

  const { loading, error, data } = UseLoading(async () =>
    fetchJSON("/api/articles")
  );

  if (loading) {
    return (
      <div>
        <div>Loading...</div>
      </div>
    );
  }
  if (error) {
    return (
      <div>
        <h1>Error</h1>
        <div>{error.toString()}</div>
      </div>
    );
  }
  return (
    <>
      <div className="mainContentContainer">
        <div className="articlesContainer">
          <ul className="article">
            {data.map((article) => {
              if (article._id === _id) {
                return (
                  <>
                    <ArticleLayoutSingle
                      key={article._id}
                      article={article}
                      user={user}
                    />
                  </>
                );
              }
            })}
          </ul>
        </div>
      </div>
    </>
  );
};

export default ArticleSingle;
