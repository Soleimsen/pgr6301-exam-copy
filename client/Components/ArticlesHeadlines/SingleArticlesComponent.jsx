import React from "react";
import { UseLoading } from "../../Hooks/UseLoading";
import { fetchJSON } from "../../Hooks/FetchJSON";
import SingleArticleLayout from "./SingleArticleLayout";

const SingleArticlesComponent = () => {
  const { loading, error, data } = UseLoading(async () =>
    fetchJSON("/api/articles")
  );

  if (loading) {
    return (
      <div>
        <div>Loading...</div>
      </div>
    );
  }
  if (error) {
    return (
      <div>
        <h1>Error</h1>
        <div>{error.toString()}</div>
      </div>
    );
  }
  return (
    <>
      <div className="articlesContainer">
        <h1>Articles</h1>
        <ul className="article">
          {data
            .slice()
            .reverse()
            .map((article) => (
              <SingleArticleLayout key={article._id} article={article} />
            ))}
        </ul>
      </div>
    </>
  );
};

export default SingleArticlesComponent;
