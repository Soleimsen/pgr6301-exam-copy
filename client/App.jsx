import React, { useContext } from "react";
import { BrowserRouter, Link, Route, Routes } from "react-router-dom";
import ArticleSingle from "./Components/ArticlesFull/ArticleSingle";
import Sidebar from "./Components/Sidebar";
import { UseLoading } from "./Hooks/UseLoading";
import { ArticlesApiContext } from "./LoginContext";
import Articles from "./Pages/Articles";
import EditArticle from "./Pages/EditArticle";
import Home from "./Pages/Home";
import Login from "./Pages/Login";
import NewArticle from "./Pages/NewArticle";
import NoPage from "./Pages/NoPage";
import Profile from "./Pages/Profile";

function UserActions({ user }) {
  let userinfo;
  if (user.google) {
    userinfo = user.google;
  }
  if (user.azure) {
    userinfo = user.azure;
  }
  if (!user || Object.keys(user).length === 0) {
    return (
      <>
        <Link className="loginBtn" to={"/login"}>
          Login
        </Link>
      </>
    );
  }

  return (
    <>
      <div className="loggedInHeader">
        <Link className="profileHeaderText" to={"/profile"}>
          {userinfo?.name ? `Profile for ${userinfo.name}` : "Profile"}
        </Link>

        <Link className="loginBtn" to={"/login/endsession"}>
          Log out
        </Link>
      </div>
    </>
  );
}

const App = () => {
  const { fetchLogin } = useContext(ArticlesApiContext);
  const { data, error, loading, reload } = UseLoading(fetchLogin);

  if (error) {
    return <div>Error: {error.toString()}</div>;
  }
  if (loading) {
    return <div>Please wait...</div>;
  }

  return (
    <div>
      <BrowserRouter>
        <header>
          <div>
            <Link className="logo" to={"/"}>
              Verdens Undergang
            </Link>
            {data.user.azure && (
              <Link className="header-article" to={"/articles/new"}>
                New Article
              </Link>
            )}
          </div>
          <div>
            <UserActions user={data?.user} />
          </div>
        </header>
        <main>
          <div className="main-container">
            <Sidebar />

            <div className="mainContentContainer">
              <Routes>
                <Route path="/" element={<Home user={data?.user} />} />
                <Route
                  path={"/profile"}
                  element={<Profile user={data?.user} />}
                />
                <Route
                  path={"/login/*"}
                  element={<Login config={data.config} reload={reload} />}
                />
                <Route path="/articles" element={<Articles />} />
                <Route
                  path="/articles/:_id"
                  element={<ArticleSingle user={data?.user} />}
                />
                <Route
                  path="/articles/new"
                  element={<NewArticle user={data?.user} />}
                />
                <Route
                  path="/articles/edit/:_id"
                  element={
                    <EditArticle user={data?.user} article={data?.article} />
                  }
                />
                <Route path="/*" element={<NoPage />} />
              </Routes>
            </div>
          </div>
        </main>
      </BrowserRouter>
    </div>
  );
};

export default App;
