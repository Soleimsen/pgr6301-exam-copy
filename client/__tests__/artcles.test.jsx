import ReactDOM from "react-dom";
import * as React from "react";
import Home from "../Pages/Home";
import Articles from "../Pages/Articles";
import NoPage from "../Pages/NoPage";
import EditArticle from "../Pages/EditArticle";
import NewArticle from "../Pages/NewArticle";
import { MemoryRouter } from "react-router-dom";
import { Simulate } from "react-dom/test-utils";
import App from "../App";
import LoginCallback from "../Pages/LoginCallback";
import Sidebar from "../Components/Sidebar";
import ArticleLayout from "../Components/ArticlesFull/ArticleLayout";
import ArticlesComponent from "../Components/ArticlesFull/ArticlesComponent";
import ArticleSingle from "../Components/ArticlesFull/ArticleSingle";
import SingleArticleLayout from "../Components/ArticlesHeadlines/SingleArticleLayout";
import SingleArticlesComponent from "../Components/ArticlesHeadlines/SingleArticlesComponent";
import SingleArticlesComponentReader from "../Components/ArticlesHeadlines/SingleArticlesComponentReader";
import LoginBtn from "../Components/Login/LoginBtn";
import StartLogin from "../Components/Login/StartLogin";
import ArticleLayoutSingle from "../Components/ArticlesFull/ArticleLayoutSingle";

describe("Pages render without crashing", () => {
  it("Articles", () => {
    const domElement = document.createElement("div");
    ReactDOM.render(<Articles />, domElement);

    expect(domElement.innerHTML).toMatchSnapshot();
  });

  it("NoPage", () => {
    const domElement = document.createElement("div");
    ReactDOM.render(<NoPage />, domElement);

    expect(domElement.innerHTML).toMatchSnapshot();
  });
  it("app", () => {
    const domElement = document.createElement("div");
    ReactDOM.render(<App />, domElement);

    expect(domElement.innerHTML).toMatchSnapshot();
  });
  it("LoginCallback", () => {
    const domElement = document.createElement("div");
    ReactDOM.render(
      <MemoryRouter>
        <LoginCallback />
      </MemoryRouter>,
      domElement
    );
    expect(domElement.innerHTML).toMatchSnapshot();
  });
});

describe("Components without crashing", () => {
  it("sidebar", () => {
    const domElement = document.createElement("div");
    ReactDOM.render(<Sidebar />, domElement);

    expect(domElement.innerHTML).toMatchSnapshot();
  });
  it("article component ", () => {
    const domElement = document.createElement("div");
    ReactDOM.render(<ArticlesComponent />, domElement);

    expect(domElement.innerHTML).toMatchSnapshot();
  });

  it("article single", () => {
    const domElement = document.createElement("div");
    ReactDOM.render(
      <MemoryRouter>
        <ArticleSingle />
      </MemoryRouter>,
      domElement
    );

    expect(domElement.innerHTML).toMatchSnapshot();
  });
  it("SingleArticlesComponent", () => {
    const domElement = document.createElement("div");
    ReactDOM.render(
      <MemoryRouter>
        <SingleArticlesComponent />
      </MemoryRouter>,
      domElement
    );

    expect(domElement.innerHTML).toMatchSnapshot();
  });
  it("SingleArticlesComponentReader", () => {
    const domElement = document.createElement("div");
    ReactDOM.render(
      <MemoryRouter>
        <SingleArticlesComponentReader />
      </MemoryRouter>,
      domElement
    );

    expect(domElement.innerHTML).toMatchSnapshot();
  });
  it("LoginBtn", () => {
    const domElement = document.createElement("div");
    ReactDOM.render(
      <MemoryRouter>
        <LoginBtn />
      </MemoryRouter>,
      domElement
    );

    expect(domElement.innerHTML).toMatchSnapshot();
  });
  it("Start login", () => {
    const domElement = document.createElement("div");
    ReactDOM.render(
      <MemoryRouter>
        <StartLogin />
      </MemoryRouter>,
      domElement
    );

    expect(domElement.innerHTML).toMatchSnapshot();
  });
});
