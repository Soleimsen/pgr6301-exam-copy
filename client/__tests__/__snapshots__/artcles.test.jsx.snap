// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components without crashing LoginBtn 1`] = `"<div><button class=\\"actualLoginBtn\\"></button></div>"`;

exports[`Components without crashing SingleArticlesComponent 1`] = `"<div><div>Loading...</div></div>"`;

exports[`Components without crashing SingleArticlesComponentReader 1`] = `"<div><div>Loading...</div></div>"`;

exports[`Components without crashing Start login 1`] = `"<div class=\\"mainContentContainer loginContainer\\"><h1 class=\\"loginTitle\\">Login</h1><div class=\\"loginDivider\\"><div><button class=\\"actualLoginBtn\\">Login with Google</button></div><div><button class=\\"actualLoginBtn\\">Login with Microsoft</button></div></div></div>"`;

exports[`Components without crashing article component  1`] = `"<div><div>Loading...</div></div>"`;

exports[`Components without crashing article single 1`] = `"<div><div>Loading...</div></div>"`;

exports[`Components without crashing sidebar 1`] = `"<div class=\\"sidebar\\"><div></div><h1>Sidebar</h1><div class=\\"topicSearch\\"><p class=\\"menuItem marginBetween\\">Topics: </p><select class=\\"menuItem\\" name=\\"Tags\\" id=\\"\\"><option value=\\"1\\" disabled=\\"\\" selected=\\"\\">Select Topic</option><option value=\\"F1\\">F1</option><option value=\\"Elon Musk\\">Elon Musk</option><option value=\\"Technology\\">Technology</option><option value=\\"Cars\\">Cars</option></select><div><button class=\\"topicBtn\\">Search topic</button></div></div></div>"`;

exports[`Pages render without crashing Articles 1`] = `"<div class=\\"mainContentContainer\\"><div><div>Loading...</div></div></div>"`;

exports[`Pages render without crashing LoginCallback 1`] = `"<h1>Please wait...</h1>"`;

exports[`Pages render without crashing NoPage 1`] = `"<div class=\\"mainContentContainer\\"><h1>404</h1><p>This page doesn't exist :(</p></div>"`;

exports[`Pages render without crashing app 1`] = `"<div>Please wait...</div>"`;
